Update CI/CD with philosophical basics (#124)
trussworks/Engineering-Playbook: developing/cicd/README.md

Here is a preferential ordering of versioning schemes:

##### Semantic Versioning

Why semver? It tells you and your customer how much has changed since the last released code and sets expectations accordingly.
If you are tagging at mainline where you are building from, you can rebuild the artifact from the same point.

From [semver.org](https://semver.org/).

```
Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes,
MINOR version when you add functionality in a backwards compatible manner, and
PATCH version when you make backwards compatible bug fixes.
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
```

##### Commit Hash

A commit hash is unique (with extremely few collisions) and is easily linked back to history in code.
However, how much has changed is opaque to your users and it is difficult to determine how old this version is in comparison to other versions.

##### Other versioning strategies

These are other versioning strategies we've seen. We do not recommend them.

> * _Feature branch related names._ These should be short lived and not maintained for more than a week at a time. Most users will not find these useful.

Reviewer: Curious how you came up with "a week".

Author: It was kinda arbitrary. I think I read something like that timeline in accelerate.... that was a long time ago.

Reviewer: I'd maybe push for "maintained for no more than a few days." A whole week sounds like an outlier to me.

* _Build id related names._ These are opaque to a user and harder to dig up history on when debugging.
* _Build date related names._ These are also opaque to a user and difficult to dig up history on when debugging. You at least get a sense of when these changes went in but are hard to tie to a commit in mainline.

### Test

We will not be writing deeply about testing methodologies or specifics around kinds of tests and the philosophies here.

In the context of a CI pipeline:

**Tests should be hermetic.** Ideally, multiple CI runs should be able to run at the same time without affecting each other.

**Tests should be idempotent.** Tests in CI should be able to run several times over without producing inconsistent effects.

**Tests should be reproducible in a developer's computer.** So developers can debug them more easily.

**Tests should run fast.** "Fast" is poorly defined here.
Generally speaking, you want tests to be fast so merging code changes is fast.

> A developer shouldn't think they can start on a new piece of work in the hour or two to push a change t hrough CI and context switch.

Reviewer: fixed typo
```suggestion
A developer shouldn't think they can start on a new piece of work in the hour or two to push a change through CI and context switch.
```

**Deployment tools should be maintained separate from the application being deployed.** We've seen this happen where we need to fix some deployed code but also had a problem with the code that performs the deployment and all of it is a hairy mess of debugging.

### Release

There are a few different ways of implementing release.
No matter what, release is the step when you allow customers to have access to that particular version of your code.

If you're releasing a CLI, it may mean that you're releasing your binary to a Homebrew tap or just a zip file to GitHub releases.

If you have a service, you could have a loadbalancer manage traffic between different versions of your services.
Or you could use something like Launch Darkly to leverage feature flags to do something similar for different user cohorts.

## Continous (Integration | Delivery | Deploy)

We throw around terms like CI or CD with the assumption that we know what the differences are between these practices.

For clarity, at Truss we'll refer to them as follows:

### Continuous Integration

`Continuous Integration` is when developers merge changes into mainline as often as possible.

> Automated build and test steps validate changes and are incorporated into review before merging to mainline.

Reviewer: Strictly speaking, CI is the practice of validating mainline on every push.  It doesn't really say anthing about branches. However, what you write is Best Practice as it means finding failure before we get to mainline. Ideally, mainline should never fail because that blocks changes from being merged. This is where a tool like [bors](https://bors.tech/) becomes interesting. It's probably worth talking about the value of keeping mainline failure free. Let me know if you want me to wax more on that.

Author: Made some edits to better reflect your views here. I think we might want more info on the value of keeping mainline failure free but I think that's maybe more detail than I was going for in this PR.

> This helps mitigate risks from long lived branches but you have automated tests and a developer workflow that allows for frequent merges.

Reviewer: This is only true if the developer is regularly merging mainline in their branch. That's really important to point out.

> Small incremental changes to code and small discrete tests facilitate freqent merges in CI.

Reviewer: It also makes every change easier to understand. I think we undervalue this and it's one of the reason I dislike squashing.

### Continuous Delivery

`Continuous Delivery` can be described as an extension to `Continuous Integration`.
In addition to automated build and test steps running on every proposed change to mainline, the delivery of the code is also automated.

You want the software built to be deployable at any given point in time.

Delivery may mean different things depending on your workflow.
It may mean that you have built your artifacts and validated them and delivered them to a centralized repository.

> For example, Docker containers might go to DockerHUB or ECR.

Reviewer: fixed name and expanded acronym
```suggestion
For example, Docker containers might go to Docker Hub or Amazon Elastic Container Registry (ECR).
```

Might mean that the code has been deployed to a staging environment.
All of these could be "delivered" depending on the project's workflow.

### Continous Deployment

> `Continuous Delivery` can be described as a further extension of `Continuous Delivery`.

Reviewer: fixed typo
```suggestion
`Continuous Deployment` can be described as a further extension of `Continuous Delivery`.
```

> In addition to automated build, test, and delivery production deployment is also automated.

Reviewer: added punctuation
```suggestion
In addition to automated build, test, and delivery, production deployment is also automated.
```

> Generally speaking, you will want to have some good logging and monitoring instrumentation so that you can automate rolling back or forward when something goes wrong.

Reviewer: I'd want to include something like …

This is an advanced state that requires excellent, trustworthy, automated tests and monitoring. The complexity of continuous deployment system is correlated with the complexity of the delivery pipeline. It's much easier to have continous deployment of a single container image than a full-feature web application.

These branches need to be kept up to date with mainline with frequent merges or rebases from mainline.

The point of either of those practices is to minimize drift between your working branch and mainline so validation in CI or manually is easier.

### Continuous Delivery

`Continuous Delivery` can be described as an extension to `Continuous Integration`.
In addition to automated build and test steps running on every proposed change to mainline, the delivery of the code is also automated.

You want the software built to be deployable at any given point in time.

Delivery may mean different things depending on your workflow.
It may mean that you have built your artifacts and validated them and delivered them to a centralized repository.
For example, Docker containers might go to Docker Hub or Amazon Elastic Container Registry (ECR).
Might mean that the code has been deployed to a staging environment.
All of these could be "delivered" depending on the project's workflow.

### Continous Deployment

`Continuous Deployment` can be described as a further extension of `Continuous Delivery`.
In addition to automated build, test, and delivery, production deployment is also automated.

> This is an advanced state that requires excellent, trustworthy, automated tests and monitoring. The complexity of continuous deployment system is correlated with the complexity of the delivery pipeline. It's much easier to have continous deployment of a single container image than a full-feature web application.

Reviewer: fix typo
```suggestion
This is an advanced state that requires excellent, trustworthy, automated tests and monitoring. The complexity of continuous deployment system is correlated with the complexity of the delivery pipeline. It's much easier to have continous deployment of a single container image than a full-featured web application.
```

### Continuous Integration

`Continuous Integration` is when you validate on every push to mainline.
This helps folks find failures in the code before it reaches mainline.
Ideally, automated build and test steps validate changes and stop any new changes to mainline.
Mainline must always be in a good state to deploy from.

This may manifest as a process where developers merge changes into mainline as often as possible.

> Small incremental changes to code and small discrete tests make changes easier to understand and facilitate freqent merges in CI.

Reviewer: fix typo
```suggestion
Small incremental changes to code and small discrete tests make changes easier to understand and facilitate frequent merges in CI.
```
